File: 1653-minimum-deletions-to-make-string-balanced/1653-minimum-deletions-to-make-string-balanced.go
func minimumDeletions(s string) int {
    n := len(s)
    stack := []byte{}
    ans := 0
    for i := 0; i < n; i++ {
        if len(stack) == 0 { stack = append(stack, s[i]); continue }

        if s[i] == 'a' && stack[len(stack)-1] == 'b' {
            ans++
            stack = stack[:len(stack)-1]
        } else {
            stack = append(stack, s[i])
        }
    }

    return ans
}

// stack, string
// time: O(n)
// space: O(n)

File: 1653-minimum-deletions-to-make-string-balanced/1653-minimum-deletions-to-make-string-balanced.py
class Solution:
    def minimumDeletions(self, s: str) -> int:
        stack = []
        ans = 0
        for i in range(0, len(s)):
            if len(stack) == 0:
                stack.append(s[i])
                continue
            
            if stack[len(stack)-1] == 'b' and s[i] == 'a':
                ans += 1
                stack.pop()
            else:
                stack.append(s[i])
        
        return ans

# stack, string
# time: O(n)
# space: O(n)